File: architecture-skills/design-patterns/examples/command/command.java
// Command pattern in Java.
//
// Text editor with undo: a Command interface with execute/undo, concrete
// commands capturing their own undo state, and a history deque as the
// invoker. For one-shot commands without undo, a Runnable lambda is the
// modern lightweight form.

import java.util.ArrayDeque;
import java.util.Deque;

public class command {

    /** Receiver. */
    static final class Editor {
        StringBuilder text = new StringBuilder();
    }

    interface Command {
        void execute();
        void undo();
    }

    static final class InsertText implements Command {
        private final Editor editor;
        private final String text;

        InsertText(Editor editor, String text) {
            this.editor = editor;
            this.text = text;
        }

        @Override
        public void execute() {
            editor.text.append(text);
        }

        @Override
        public void undo() {
            editor.text.setLength(editor.text.length() - text.length());
        }
    }

    static final class DeleteLast implements Command {
        private final Editor editor;
        private final int count;
        private String deleted; // captured on execute so undo can restore it

        DeleteLast(Editor editor, int count) {
            this.editor = editor;
            this.count = count;
        }

        @Override
        public void execute() {
            int len = editor.text.length();
            deleted = editor.text.substring(len - count);
            editor.text.setLength(len - count);
        }

        @Override
        public void undo() {
            editor.text.append(deleted);
        }
    }

    /** Invoker: runs commands and tracks them for undo. */
    static final class History {
        private final Deque<Command> undoStack = new ArrayDeque<>();

        void execute(Command command) {
            command.execute();
            undoStack.push(command);
        }

        void undo() {
            if (!undoStack.isEmpty()) {
                undoStack.pop().undo();
            }
        }
    }

    public static void main(String[] args) {
        Editor editor = new Editor();
        History history = new History();

        history.execute(new InsertText(editor, "Hello"));
        history.execute(new InsertText(editor, ", world!"));
        history.execute(new DeleteLast(editor, 1));
        System.out.println("after edits: \"" + editor.text + "\"");

        history.undo();
        history.undo();
        System.out.println("after 2 undos: \"" + editor.text + "\"");
    }
}

File: architecture-skills/design-patterns/examples/command/command.swift
// Command pattern in Swift.
//
// Text editor with undo: a protocol with execute/undo, concrete commands
// capturing their own undo state, and a history array as the invoker.
// Foundation's UndoManager is the platform's built-in invoker; closures
// cover one-shot commands that need no undo.

final class Editor { // receiver
    var text = ""
}

protocol Command {
    func execute()
    func undo()
}

final class InsertText: Command {
    private let editor: Editor
    private let text: String

    init(editor: Editor, text: String) {
        self.editor = editor
        self.text = text
    }

    func execute() {
        editor.text += text
    }

    func undo() {
        editor.text.removeLast(text.count)
    }
}

final class DeleteLast: Command {
    private let editor: Editor
    private let count: Int
    private var deleted = "" // captured on execute so undo can restore it

    init(editor: Editor, count: Int) {
        self.editor = editor
        self.count = count
    }

    func execute() {
        deleted = String(editor.text.suffix(count))
        editor.text.removeLast(count)
    }

    func undo() {
        editor.text += deleted
    }
}

final class History { // invoker
    private var undoStack: [Command] = []

    func execute(_ command: Command) {
        command.execute()
        undoStack.append(command)
    }

    func undo() {
        undoStack.popLast()?.undo()
    }
}

let editor = Editor()
let history = History()

history.execute(InsertText(editor: editor, text: "Hello"))
history.execute(InsertText(editor: editor, text: ", world!"))
history.execute(DeleteLast(editor: editor, count: 1))
print("after edits: \"\(editor.text)\"")

history.undo()
history.undo()
print("after 2 undos: \"\(editor.text)\"")
